pkg/gardenlet/controller/shoot/shoot_control.go

```diff
@@ -243,7 +243,7 @@ func shouldPrepareShootForMigration(shoot *gardencorev1beta1.Shoot) bool {
 const taskID = "initializeOperation"
 
 func (c *Controller) initializeOperation(ctx context.Context, logger *logrus.Entry, gardenClient kubernetes.Interface, shoot *gardencorev1beta1.Shoot, project *gardencorev1beta1.Project, cloudProfile *gardencorev1beta1.CloudProfile, seed *gardencorev1beta1.Seed) (*operation.Operation, error) {
-	gardenSecrets, err := garden.ReadGardenSecrets(ctx, gardenClient.Cache(), c.seedLister, gutil.ComputeGardenNamespace(seed.Name))
+	gardenSecrets, err := garden.ReadGardenSecrets(ctx, gardenClient.Client(), c.seedLister, gutil.ComputeGardenNamespace(seed.Name))
```

Comment:
I don't know/can't judge if it really saves that much to justify the (yet again (we already have this (unused?) multi-namespace cache)) additional complexity. @timebertt WDYT? I can do the change if you think it's required, of course.

Reply:
I'm confused. The garden client has a secrets watch for the seed namespace, see here:
https://github.com/gardener/gardener/blob/453b70ed9a40c0eb72694452d55d54b9d074ea29/pkg/client/kubernetes/clientmap/internal/garden_clientmap.go#L81
Why do you think it is unused?

Also, what was the reasoning behind changing from `Cache()` to `Client()`?

Reply:
@timebertt When I use `Cache()` instead of `Client()` then a cluster-wide watch for `Secret`s is trying to be established which is not allowed:

```
E0520 07:57:10.884194   75691 reflector.go:138] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:229: Failed to watch *v1.Secret: failed to list *v1.Secret: secrets is forbidden: User "gardener.cloud:system:seed:foo" cannot list resource "secrets" in API group "" at the cluster scope: only the following verbs are allowed for this resource type: [create get patch update delete]
```

Hence, I switched to `Client()` for now. Do you have other recommendations?

Reply:
Ah, that's only the case in your local setup, where you have the `seedSelector` set, is that right?
In that case, the namespace-specific cache, which I linked above, is skipped. And then the `Cache()` opens a cluster-wide watch. Ok, understood now...
I think the switch is fine then, as the namespace-specific watch will then be used once we have graduated the `CachedRuntimeClients` feature gate. Also, once we do https://github.com/gardener/gardener/issues/4078, the `seedSelector` case will also be gone.

Coming back to your question:

> I don't know/can't judge if it really saves that much to justify the (yet again (we already have this (unused?) multi-namespace cache)) additional complexity. @timebertt WDYT?

I don't know, if we can really save so much network traffic / load and if that's really worth it. The calls here are only done exactly once per reconciliation/deletion. The care controller already does the "caching", here: https://github.com/gardener/gardener/blob/95b86ca115e46bb215d5a0793fccf69c3ff27398/pkg/gardenlet/controller/shoot/shoot_care_control.go#L274
But there it's much more critical, as this is executed once per care operation.

Reply:
@timebertt Thanks for your assessment!
@timuthy Are you then okay with keeping it as is for now? We can adapt this if we see the need, of course. In parallel, the `CachedRuntimeClient` story is progressing and will eventually relieve us from such concerns automatically. WDYT?

Reply:
> I don't know/can't judge if it really saves that much

> I don't know, if we can really save so much network traffic

Since we all don't know, I guess the way to go is to try it out 😄 🤞